pkg/kubelet/kubeletconfig/checkpoint/download.go

```diff
@@ -46,16 +57,34 @@ type Payload interface {
 
 // RemoteConfigSource represents a remote config source object that can be downloaded as a Checkpoint
 type RemoteConfigSource interface {
-	// UID returns a globally unique identifier of the source described by the remote config source object
-	UID() string
 	// KubeletFilename returns the name of the Kubelet config file as it should appear in the keys of Payload.Files()
 	KubeletFilename() string
+
 	// APIPath returns the API path to the remote resource, e.g. its SelfLink
 	APIPath() string
-	// Download downloads the remote config source object returns a Payload backed by the object,
-	// or a sanitized failure reason and error if the download fails
-	Download(client clientset.Interface) (Payload, string, error)
-	// Encode returns a []byte representation of the NodeConfigSource behind the RemoteConfigSource
+
+	// UID returns a globally unique (space and time) identifier for the payload targeted by the source.
+	// This value is updated when Download is called.
+	UID() string
+
+	// ResourceVersion returns a resource version for the payload targeted by the source.
+	// This value is updated when Download is called.
+	ResourceVersion() string
+
+	// Download downloads the remote config source's target object and returns a Payload backed by the object,
+	// or a sanitized failure reason and error if the download fails.
+	// Download takes an optional store as an argument. If provided, Download will check this store for the
+	// target object prior to contacting the API server.
+	// Download internally sets the UID and ResourceVersion on the source, based on the downloaded payload.
```

Comment:
nit: similar to above about describing internal behavior

Reply:
Actually, I think `Payload()` may be a better interface.  The fact that it downloads and caches it on disk can be an implementation detail?

Reply:
Scratch that.  Downloading the payload locally is an important part of the function.